zulipterminal/model.py

```diff
@@ -383,11 +383,41 @@ def update_reaction(self, response: Dict[str, Any]) -> None:
             self.index['messages'][message_id] = message
             self.update_rendered_view(message_id)
 
+    def update_read_status(self, event: Dict[str, Any]) -> None:
+        assert not event['all']  # FIXME Should handle
+        indexed_messages = set(self.index['messages'])
+        messages_to_mark_read = set(event['messages'])
+
+        for message_id in messages_to_mark_read & indexed_messages:
+            msg = self.index['messages'][message_id]
+            if event['operation'] == 'add':
+                if 'read' not in msg['flags']:
+                    msg['flags'].append('read')
+            elif event['operation'] == 'remove':
+                if 'read' in msg['flags']:
+                    msg['flags'].remove('read')
+            else:
+                raise RuntimeError(event, msg['flags'])
+
+            self.index['messages'][message_id] = msg
+            self.update_rendered_view(message_id)
+
+        # set_count depends on message being in the index
+        for message_id in messages_to_mark_read - indexed_messages:
+            self.get_messages(num_after=0, num_before=0, anchor=message_id)
```

Comment:
Generally, it's not great to do potentially dozens/hundreds of API queries to the server like this; for one thing, you'll hit server rate limits.  I think the correct thing to do here for marking messages as unread would be potentially to follow the webapp's system of having ultra-thin object for unread count tracking only for messages you know exist and are unread but don't have a copy of.  

The other option would be to add a bulk version of this query (`get_messages_by_id`, e.g.), but that seems wrong since the client doesn't actually need a heavy version of those messages.  Note that it's very possible to mark 1000+ messages as read all at once.

Reply:
@neiljp we are already tracking unread counts for all the messages (received or not in zt) via `model.unread_counts` . We can update the new counts in `model.unread_counts` and update them accordings via `set_count` in zt. I also don't think we need to do these many queries to the server.

Reply:
Is this the error caused by the rate limit or something else?
```
Traceback (most recent call last):
  File "/home/rht/code/venv/lib/python3.7/site-packages/zulip_term-0.2.1+git-py3.7.egg/zulipterminal/cli/run.py", line 146, in main
    Controller(zuliprc_path, THEMES[theme_to_use[0]]).main()
  File "/home/rht/code/venv/lib/python3.7/site-packages/zulip_term-0.2.1+git-py3.7.egg/zulipterminal/core.py", line 36, in __init__
    self.view = View(self)
  File "/home/rht/code/venv/lib/python3.7/site-packages/zulip_term-0.2.1+git-py3.7.egg/zulipterminal/ui.py", line 35, in __init__
    super(View, self).__init__(self.main_window())
  File "/home/rht/code/venv/lib/python3.7/site-packages/zulip_term-0.2.1+git-py3.7.egg/zulipterminal/ui.py", line 90, in main_window
    self.center_column = self.message_view()
  File "/home/rht/code/venv/lib/python3.7/site-packages/zulip_term-0.2.1+git-py3.7.egg/zulipterminal/ui.py", line 43, in message_view
    self.search_box)
  File "/home/rht/code/venv/lib/python3.7/site-packages/zulip_term-0.2.1+git-py3.7.egg/zulipterminal/ui_tools/views.py", line 278, in __init__
    msg_list = MessageView(model)
  File "/home/rht/code/venv/lib/python3.7/site-packages/zulip_term-0.2.1+git-py3.7.egg/zulipterminal/ui_tools/views.py", line 35, in __init__
    self.set_focus(self.focus_msg)
  File "/home/rht/code/venv/lib/python3.7/site-packages/urwid/listbox.py", line 581, in set_focus
    self._body.set_focus(position)
  File "/home/rht/code/venv/lib/python3.7/site-packages/urwid/listbox.py", line 238, in set_focus
    self.focus = position
  File "/home/rht/code/venv/lib/python3.7/site-packages/urwid/monitored_list.py", line 165, in _set_focus
    raise IndexError('focus index is out of range: %s' % (index,))
IndexError: focus index is out of range: 10000000000000000
Zulip Terminal has crashed!
You can ask for help at:
https://chat.zulip.org/#narrow/stream/206-zulip-terminal
```

Reply:
@rht this is different issue which is fixed on master.

Reply:
@amanagr @timabbott This was intended as a v1 towards this approach, and I agree we can work on a simplifier approach which only stores the unread flags. I'll look at that for 0.3.1+ I think now?

Reply:
I think creating an issue for it so that it doesn't lose track would be nice.  (Also commenting it in the code itself would be great if are doing it in a later commit.)